internal/controller/linodefirewall_controller_helpers.go

```diff
@@ -313,13 +314,26 @@ func processAddresses(addresses *infrav1alpha2.NetworkAddresses) (ipv4s, ipv6s [
 	return slices.Collect(maps.Keys(ipv4Set)), slices.Collect(maps.Keys(ipv6Set))
 }
 
-// get finalizer for Firewall that references the AddressSet
+// get finalizer for Firewall that references the AddressSet or FirewallRule
 func getFinalizer(lfw *infrav1alpha2.LinodeFirewall) string {
 	return fmt.Sprintf("lfw.%s.%s", lfw.Namespace, lfw.Name)
 }
 
+func addFinalizer(ctx context.Context, k8sClient clients.K8sClient, lfw *infrav1alpha2.LinodeFirewall, obj client.Object) error {
+	finalizer := getFinalizer(lfw)
+	if !controllerutil.ContainsFinalizer(obj, finalizer) {
+		controllerutil.AddFinalizer(obj, finalizer)
+		if retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			return k8sClient.Update(ctx, obj)
+		}); retryErr != nil {
+			return retryErr
+		}
```

Comment:
I'm not sure if this retry logic solves the race condition from trying to manage the same object concurrently [[1]].

From the [GoDoc](https://pkg.go.dev/k8s.io/client-go/util/retry#RetryOnConflict):

>  RetryOnConflict is used to make an update to a resource when you have to worry about conflicts caused by other code making unrelated updates to the resource at the same time. **fn should fetch the resource to be modified, make appropriate changes to it, try to update it, and return (unmodified) the error from the update function.**

I think you'll just be retrying the update using the same `client.Object`(with the same stale resource version [[2]]) here. Maybe you could also try a patch ([[3]], [[4]]) instead of retrying?

[1]: https://book.kubebuilder.io/reference/good-practices#why-is-it-recommended-to-avoid-a-scenario-where-multiple-controllers-are-updating-the-same-custom-resource-cr
[2]: https://alenkacz.medium.com/kubernetes-operators-best-practices-understanding-conflict-errors-d05353dff421
[3]: https://kubernetes.io/docs/reference/using-api/api-concepts/#api-verbs
[4]: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#Patch

Reply:
![slap](https://github.com/user-attachments/assets/98c67b1f-e6fc-4b4d-ac69-2ec6f9918e5c)